### PR TITLE
Use 'never' type for empty array literals

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR introduces a widening form of the `never` type that we use for empty array literals in `--strictNullChecks` mode. This fixes the following issue reported in #8878:

```typescript
// Compile with --strictNullChecks
function concat<T>(xs: T[], ys: T[]): T[] {
    return [...xs, ...ys];
}
const y = concat([], ["a"]);  // Should be type string[]
```

Before this fix the inferred type above was `(string | undefined)[]` because we were using type `undefined[]` for empty array literals, and `undefined` isn't a bottom type in strict null checking mode.

### CODE REVIEW

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -119,9 +119,9 @@ namespace ts {
         const voidType = createIntrinsicType(TypeFlags.Void, "void");
         const undefinedType = createIntrinsicType(TypeFlags.Undefined | nullableWideningFlags, "undefined");
         const nullType = createIntrinsicType(TypeFlags.Null | nullableWideningFlags, "null");
-        const emptyArrayElementType = createIntrinsicType(TypeFlags.Undefined | TypeFlags.ContainsUndefinedOrNull, "undefined");
         const unknownType = createIntrinsicType(TypeFlags.Any, "unknown");
         const neverType = createIntrinsicType(TypeFlags.Never, "never");
+        const wideningNeverType = createIntrinsicType(TypeFlags.Never | TypeFlags.ContainsWideningType, "never");
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Leave a comment on what a *widening* never implies over an unwidened never. From what I understand, it's kind of a similar concept to freshness.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9187,7 +9187,8 @@ namespace ts {
                     }
                 }
             }
-            return createArrayType(elementTypes.length ? getUnionType(elementTypes) : emptyArrayElementType);
+            const elementType = getUnionType(elementTypes);
+            return createArrayType(elementType === neverType ? strictNullChecks ? wideningNeverType : undefinedType : elementType);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: So if you have an array of `[a, b, c]`, and `a`, `b`, and `c` each have type `never`, does the type of the array literal get widened to `any[]`?

Comment (ahejlsberg) [inline src/compiler/checker.ts]: Yes, the array literal would widen to `any[]`. We could do extra work to aggregate widening and only widen only if all (or if just one?) of the elements are of the widening `never` type. But it hardly seems worth it.

**Comment by ahejlsberg** [comment]
Superceded by #8944.

**Pull request closed without merging**